### PR TITLE
HTCONDOR-772 Implement checkpoint_destination

### PR DESCRIPTION
**Pull request by Todd-L-Miller**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [x] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_shadow.V6.1/remoteresource.cpp [hunk 1]
```diff
@@ -2176,6 +2189,96 @@ RemoteResource::initFileTransfer()
...
+	if( largestManifestNumber == -1 ) {
+		return;
+	}
+
+	// Validate the candidate manifests in reverse ordinal order.
+	int manifestNumber = -1;
+	std::string manifestFileName;
+	if( largestManifestNumber != -1 ) {
+		for( int i = largestManifestNumber; i >= 0; --i ) {
+			formatstr( manifestFileName, "%s/MANIFEST.\%.4d", spoolPath.c_str(), i );
```
Comment (JaimeFrey) [inline src/condor_shadow.V6.1/remoteresource.cpp:2228]:
```suggestion
			formatstr( manifestFileName, "%s/MANIFEST.%.4d", spoolPath.c_str(), i );
```

File: src/condor_utils/file_transfer.cpp [hunk 2]
```diff
@@ -256,6 +257,17 @@ class FileTransferItem {
...
 
+void
+dPrintFileTransferList( int flags, const FileTransferList & list, const std::string & header ) {
+	std::string message = header;
+	for( auto & i: list ) {
+		formatstr_cat( message, " %s -> %s [%s],",
+			i.srcName().c_str(), i.destDir().c_str(), i.destUrl().c_str()
+		);
+	}
+	dprintf( flags, "%s\n", message.c_str() );
```
Comment (JaimeFrey) [inline src/condor_utils/file_transfer.cpp:268]: Would be nice to not have a trailing comma at the end of the log message.

File: src/condor_utils/file_transfer.h [hunk 3]
```diff
@@ -387,6 +417,43 @@ class FileTransfer final: public Service {
...
 	int DoUpload( filesize_t *total_bytes, ReliSock *s);
+	int DoCheckpointUploadFromStarter( filesize_t * total_bytes, ReliSock * s );
+	int DoCheckpointUploadFromShadow( filesize_t * total_bytes, ReliSock * s );
+	int DoNormalUpload( filesize_t * total_bytes, ReliSock * s );
+
+	typedef struct {
+		filesize_t peer_max_transfer_bytes = {-1};
+		bool I_go_ahead_always = {false};
+		bool peer_goes_ahead_always = {false};
+		bool socket_default_crypto;
```
Comment (JaimeFrey) [inline src/condor_utils/file_transfer.h:428]:
```suggestion
		bool socket_default_crypto = {true};
```

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
* Use full pathnames when accessing the MANIFEST file. You can use FileTransfer::Iwd, JobInfoCommunicator::job_iwd, or Starter::GetWorkingDir(false) in the starter.
* CheckpointNumber in the job ad appears to be the id of the next checkpoint to take. Shouldn't it be the id of the most recent one taken?
* Could we obscure the MANIFEST filename (e.g. with "_condor_" prefix) to minimize chances of conflicting the job's filenames?

File: src/condor_utils/file_transfer.cpp (hunk 2 above)
Comment (Todd-L-Miller) [reply src/condor_utils/file_transfer.cpp]: Done.

File: src/condor_utils/file_transfer.cpp [hunk 4]
```diff
@@ -265,6 +265,8 @@ dPrintFileTransferList( int flags, const FileTransferList & list, const std::str
 			i.srcName().c_str(), i.destDir().c_str(), i.destUrl().c_str()
 		);
 	}
+	// Don't print the trailing comma.
+	message[message.size()] = '\0';
```
Comment (JaimeFrey) [inline src/condor_utils/file_transfer.cpp:269]: Three problems with this line of code:
* It doesn't do anything (you wanted size()-1)
* You should be shortening the string (e.g. with erase()), not overwriting characters with NUL.
* If the list is empty, you erase part of the header string.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
Fix removal of trailing comma.

File: src/condor_utils/file_transfer.cpp (hunk 4 above)
Comment (Todd-L-Miller) [reply src/condor_utils/file_transfer.cpp]: Strictly speaking, none of this code does anything because all of the calls are commented out, but fine. :)

**Comment by Todd-L-Miller** [comment]
In `RemoteResource::initFileTransfer()`, each manifest file is accessed via the path-to-spool, which is an absolute path.

In the test, the manifest file is checked accessed via the `test_dir`, which is an absolute path.

In `createCheckpointManifest()` the files whose checksums are being computed are passed in via relative path, and we have to record them as relative paths; as far as I can tell, writing the corresponding record to an absolute path only adds complication to the code and the possibility of error.

The only other place the manifest file(s) are mentioned by name, they are not being read from.

**Comment by JaimeFrey** [comment]
My concern for the relative path was in JICShadow::transferCompleted() and FileTransfer::createCheckpointManifest(), where we assume we're in the job's scratch directory when handling the manifest file. It appears the starter is very deliberate about doing that. That makes me wonder where starter core files end up.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
The  code looks fine now. But the manifest test is failing.

**Comment by Todd-L-Miller** [comment]
> But the manifest test is failing.

How did _that_ happen?  Thanks for the heads-up.

**Comment by Todd-L-Miller** [comment]
> How did that happen?
Because I forgot to update the test when I changed the pattern for legal MANIFEST file names.  Should be all good now.

**Review by JaimeFrey — APPROVED** [review]
Looks good.

**Pull request merged**